src/Sylius/Bundle/ApiBundle/Doctrine/QueryCollectionExtension/ExchangeRateCollectionExtension.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Doctrine\QueryCollectionExtension;
+
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Extension\ContextAwareQueryCollectionExtensionInterface;
+use ApiPlatform\Core\Bridge\Doctrine\Orm\Util\QueryNameGeneratorInterface;
+use Doctrine\ORM\QueryBuilder;
+use Sylius\Bundle\ApiBundle\Context\UserContextInterface;
+use Sylius\Bundle\ApiBundle\Serializer\ContextKeys;
+use Sylius\Component\Core\Model\AdminUserInterface;
+use Sylius\Component\Currency\Model\ExchangeRate;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class ExchangeRateCollectionExtension implements ContextAwareQueryCollectionExtensionInterface
+{
+    public function __construct(private UserContextInterface $userContext)
+    {
+    }
+
+    public function applyToCollection(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+        string $resourceClass,
+        string $operationName = null,
+        array $context = []
+    ): void {
+        if (!is_a($resourceClass, ExchangeRate::class, true)) {
+            return;
+        }
+
+        $user = $this->userContext->getUser();
+        if ($user instanceof AdminUserInterface && in_array('ROLE_API_ACCESS', $user->getRoles(), true)) {
+            return;
+        }
+
+        Assert::keyExists($context, ContextKeys::CHANNEL);
+        $channel = $context[ContextKeys::CHANNEL];
+
+        $currencyParameterName = $queryNameGenerator->generateParameterName('currency');
+        $rootAlias = $queryBuilder->getRootAliases()[0];
+
+        $queryBuilder
+            ->andWhere(sprintf('%s.sourceCurrency = :%s', $rootAlias, $currencyParameterName))
```

Comment:
If I'm correct, you should filter here if the `baseCurrency` is the `sourceCurrency` or `targetCurrency`, and cover that in the scenario and contract test